runtime/parachains/src/scheduler.rs

```diff
@@ -13,3 +13,1403 @@
+		/// Currently scheduled cores - free but up to be occupied. Ephemeral storage item that's wiped on finalization.
+		///
+		/// Bounded by the number of cores: one for each parachain and parathread multiplexer.
+		Scheduled get(fn scheduled): Vec<CoreAssignment>; // sorted ascending by CoreIndex.
+	}
+}
+
+decl_error! {
+	pub enum Error for Module<T: Trait> { }
+}
+
+decl_module! {
+	/// The scheduler module.
+	pub struct Module<T: Trait> for enum Call where origin: <T as system::Trait>::Origin {
+		type Error = Error<T>;
+	}
+}
+
+impl<T: Trait> Module<T> {
+	/// Called by the initializer to initialize the scheduler module.
+	pub(crate) fn initializer_initialize(_now: T::BlockNumber) -> Weight {
+		Self::schedule(Vec::new());
+
+		0
+	}
+
+	/// Called by the initializer to finalize the scheduler module.
+	pub(crate) fn initializer_finalize() {
+		// Free all scheduled cores and return parathread claims to queue, with retries incremented.
+		let config = <configuration::Module<T>>::config();
+		ParathreadQueue::mutate(|queue| {
+			for core_assignment in Scheduled::take() {
+				if let AssignmentKind::Parathread(collator, retries) = core_assignment.kind {
+					let entry = ParathreadEntry {
+						claim: ParathreadClaim(core_assignment.para_id, collator),
+						retries: retries + 1,
+					};
+
+					if entry.retries > config.parathread_retries { continue }
+					queue.queue_entry(entry, config.parathread_cores);
+				}
+			}
+		})
+
+	}
+
+	/// Called by the initializer to note that a new session has started.
+	pub(crate) fn initializer_on_new_session(notification: &SessionChangeNotification<T::BlockNumber>) {
+		let &SessionChangeNotification {
+			ref validators,
+			ref random_seed,
+			ref new_config,
+			..
+		} = notification;
+		let config = new_config;
+
+		let mut thread_queue = ParathreadQueue::get();
+		let n_parachains = <paras::Module<T>>::parachains().len() as u32;
+		let n_cores = n_parachains + config.parathread_cores;
+
+		<SessionStartBlock<T>>::set(<system::Module<T>>::block_number());
+		AvailabilityCores::mutate(|cores| {
+			// clear all occupied cores.
+			for maybe_occupied in cores.iter_mut() {
+				if let Some(CoreOccupied::Parathread(claim)) = maybe_occupied.take() {
+					let queued = QueuedParathread {
+						claim,
+						core_offset: 0, // this gets set later in the re-balancing.
+					};
+
+					thread_queue.queue.push(queued);
+				}
+			}
+
+			cores.resize(n_cores as _, None);
+		});
+
+		// shuffle validators into groups.
+		if n_cores == 0 || validators.is_empty() {
+			ValidatorGroups::set(Vec::new());
+		} else {
+			let mut rng: ChaCha20Rng = SeedableRng::from_seed(*random_seed);
+
+			let mut shuffled_indices: Vec<_> = (0..validators.len())
+				.enumerate()
+				.map(|(i, _)| i as ValidatorIndex)
+				.collect();
+
+			shuffled_indices.shuffle(&mut rng);
+
+			let group_base_size = validators.len() / n_cores as usize;
+			let larger_groups = validators.len() % n_cores as usize;
```

Comment:
^^^ really is the number of larger groups

```suggestion
			let n_larger_groups = validators.len() % n_cores as usize;
```

for consistency